core/src/main/java/org/apache/calcite/jdbc/Driver.java

```diff
@@ -64,6 +64,15 @@ public Driver() {
     this.prepareFactory = createPrepareFactory();
   }
 
+  private Driver(Function0<CalcitePrepare> prepareFactory) {
```

Comment:
unify constructors. all constructors except one should call other constructors.

Reply:
I've unified the constructors in my last commit: https://github.com/apache/calcite/pull/3084/commits/5314d5d29e5cd6d271626d7b877125278e8292d4#diff-202586edbf28484a3a4c1498d8f61096e7a12c0a9c813260fb4282b5c19a3e96R63

But it doesn't compile due to no explicit assigning of prepareFactory. 
Do you have suggestions here?

Reply:
You're right there's a problem. The problem is that the constructor calls `createPrepareFactory()` - an instance method - before the constructor is complete, and that's a code smell, because it doesn't allow the class to be subclassed cleanly.

Is there a way to change the design of `Driver` so that it doesn't have that code smell?

I like the idea of adding `Driver withPrepareFactory(Supplier<CalcitePrepare>)` so that people don't have to subclass Driver. But we have to do so compatibly.

Reply:
The way that its designed seems to imply some conflicting restraints
- we need `createPrepareFactory()` to be override-able (use cases in JdbcTest.java), thus has to stay an instance method 
- we want `prepareFactory` to be `final` 
- all constructors except 1 have to call 1 constructor

Is there a way to satisfy all 3 without conflicts?

Reply:
Pushed a change so that `private Driver(prepareFactory)` calls the `public Driver()`